### DISCUSSION TITLE
Best practise to build a ToC

### DISCUSSION THREAD

**Discussion by ghost** [thephpleague/commonmark#649] [Q&A] [answered]
Hi,

I'm wondering what is the best way to build a table of contents for a page.
I've replaced the `HeadingRenderer` with a custom one to add a `<a name="title-slugified">` link inside the heading. That was rather easy. Now, I need to build the ToC itself. I see two different approaches:
1. Walk through the AST
2. Use the block renderer (or maybe a custom block parser?)

The first one seems a better practise to me, since it separates concerns, while the latter seems more efficient, since there is no need to go through the document again: when a heading is rendered, I know it's a heading, I can handle it too and add it to the ToC (which would be injected in the `HeadingRenderer`'s constructor).

What do you think?

I'm thinking of a third method, but it requires features the library currently does not offer: events triggered by parsers and renderers. In my case, I've overridden the `HeadingRenderer`, which is not a big deal, because it's a rather small class, but is not the best practise (I think). Listening to events would allow us to just wrap the content of a block in another tag, add an attribute, store some information while parsing, etc.

**Comment by colinodell** [comment]
Currently, I think walking the AST is probably the best option.  I'm definitely interested in expanding the usage of events in this library - if you'd be open to sharing your ideas that would help me better understand how to best implement that :)

**Comment by ghost** [comment]
Thanks for your answer.

I'll open a new issue about events in the next few days.

**Comment by colinodell** [comment, marked as answer]
For anyone who comes across this question - we do have a [Table of Contents extension](https://commonmark.thephpleague.com/extensions/table-of-contents/) now!
